### ISSUE TITLE
React Optimized Components need proper parents

### ISSUE THREAD

**Issue by cblappert** [facebook/prepack#2550]
Normal optimized functions have a `parentAdditionalFunction` field in their `AdditionalFunctionEffects` that points to the outer optimized function with regards to nesting of optimized functions.

This parent chain is used to ensure that code is emitted to the proper locations when using nested optimized functions.

**Comment by trueadm** [comment]
Please can you add some background as to what assigns `parentAdditionalFunction` and why things work now without it? It would be good for someone on the team other than I to take a stab at this – purely to get some understand and grasps of the React part of the codebase. However, I'm happy to take it on if not. :)

**Comment by cblappert** [comment]
I ended up fixing it in the PR because Nikolai requested it, the reason it matters now is that parentAdditionalFunction is used in two places:
1. `Generator.fromEffects` to determine what ModifiedBindings belong to environments created by this optimized function (or nested functions) so that we can elide these modifications.
2. In `tryGetOutermostOptimizedFunction` in `ResidualOptimizedFunctions`. This is used in the serializer and referentializer to determine what scope to emit things into. Previously in a bunch of places we defaulted to emitting things to global scope if we couldn't find a narrower common scope between the set of scopes associated to a value by the ResidualHeapVisitor. Now we correctly try to emit to an optimized function or global.

The only time we would produce incorrect code before was in situations like this: #2430 or #2428 which I guess were pretty rare in react compiler use cases.
